src/lib/pubkey/frodokem/frodokem_common/frodo_matrix.cpp

```diff
@@ -0,0 +1,494 @@
+
+   BOTAN_ASSERT_NOMSG(in.size() * 8 == constants.n_bar() * constants.n_bar() * constants.b());
+
+   size_t pos = 0;
+   for(size_t i = 0; i < (constants.n_bar() * constants.n_bar()) / 8; ++i) {
+      uint64_t temp = 0;
+      for(size_t j = 0; j < constants.b(); ++j) {
+         temp |= static_cast<uint64_t /* avoiding integral promotion */>(in[i * constants.b() + j]) << (8 * j);
+      }
+      for(size_t j = 0; j < 8; ++j) {
+         elements.at(pos++) = static_cast<uint16_t>((temp & mask) << (constants.d() - constants.b()));  // k*2^(D-B)
+         temp >>= constants.b();
+      }
+   }
+
+   return FrodoMatrix(dimensions, std::move(elements));
+}
+
+FrodoMatrix FrodoMatrix::add(const FrodoKEMConstants& constants, const FrodoMatrix& a, const FrodoMatrix& b) {
+   // Addition is defined for n_bar x n_bar matrices only
+   BOTAN_ASSERT_NOMSG(a.dimensions() == b.dimensions());
+   BOTAN_ASSERT_NOMSG(std::get<0>(a.dimensions()) == constants.n_bar() &&
+                      std::get<1>(a.dimensions()) == constants.n_bar());
+
+   auto elements = make_elements_vector(a.dimensions());
+
+   for(size_t i = 0; i < constants.n_bar() * constants.n_bar(); ++i) {
+      elements.at(i) = a.elements_at(i) + b.elements_at(i);
+   }
+
+   return FrodoMatrix(a.dimensions(), std::move(elements));
+}
+
+FrodoMatrix FrodoMatrix::sub(const FrodoKEMConstants& constants, const FrodoMatrix& a, const FrodoMatrix& b) {
+   // Subtraction is defined for n_bar x n_bar matrices only
+   BOTAN_ASSERT_NOMSG(a.dimensions() == b.dimensions());
+   BOTAN_ASSERT_NOMSG(std::get<0>(a.dimensions()) == constants.n_bar() &&
+                      std::get<1>(a.dimensions()) == constants.n_bar());
+
+   auto elements = make_elements_vector(a.dimensions());
+
+   for(size_t i = 0; i < constants.n_bar() * constants.n_bar(); ++i) {
+      elements.at(i) = a.elements_at(i) - b.elements_at(i);
+   }
+
+   return FrodoMatrix(a.dimensions(), std::move(elements));
+}
+
+CT::Mask<uint8_t> FrodoMatrix::constant_time_compare(const FrodoMatrix& other) const {
+   BOTAN_ASSERT_NOMSG(dimensions() == other.dimensions());
+   // TODO: Possibly use range-based comparison after #3715 is merged
+   return CT::is_equal(reinterpret_cast<const uint8_t*>(m_elements.data()),
+                       reinterpret_cast<const uint8_t*>(other.m_elements.data()),
+                       sizeof(decltype(m_elements)::value_type) * m_elements.size());
+}
+
+FrodoMatrix FrodoMatrix::mul_bs(const FrodoKEMConstants& constants, const FrodoMatrix& b, const FrodoMatrix& s) {
+   Dimensions dimensions = {constants.n_bar(), constants.n_bar()};
+   auto elements = make_elements_vector(dimensions);
+
+   for(size_t i = 0; i < constants.n_bar(); ++i) {
+      for(size_t j = 0; j < constants.n_bar(); ++j) {
+         auto& current = elements.at(i * constants.n_bar() + j);
+         current = 0;
+         for(size_t k = 0; k < constants.n(); ++k) {
+            current += static_cast<uint16_t>(
```

Comment:
This expression is very hard to read thanks to how the inner comments get formatted. Splitting it out to a couple of variables with leading comments would probably help a lot, eg

```
const uint32_t b_ink = b.elements_at(...);
// Since ...
const uint32_t s_jnk = s.elements_at(...);
current += ...
```